From 6cce2841567b7f0fd39992e2da86732865ef1b78 Mon Sep 17 00:00:00 2001
From: Morian Sonnet <MorianSonnet@googlemail.com>
Date: Thu, 9 Apr 2020 22:24:40 +0200
Subject: [PATCH] shiboken2: Activate debug Logging Categories

---
 sources/shiboken2/generator/main.cpp | 1 +
 1 file changed, 1 insertion(+)

diff --git a/sources/shiboken2/generator/main.cpp b/sources/shiboken2/generator/main.cpp
index 25daea99..891b3e7d 100644
--- a/sources/shiboken2/generator/main.cpp
+++ b/sources/shiboken2/generator/main.cpp
@@ -387,6 +387,7 @@ int main(int argc, char *argv[])
     // needed by qxmlpatterns
     QCoreApplication app(argc, argv);
     ReportHandler::install();
+    QLoggingCategory::setFilterRules(QStringLiteral("*.debug = true"));
     qCDebug(lcShiboken()).noquote().nospace() << QCoreApplication::arguments().join(QLatin1Char(' '));
 
     // Store command arguments in a map
-- 
2.25.2

